### ISSUE TITLE
How to apply the model on customised dataset

### ISSUE THREAD

**Issue by Jacobieee** [YeongHyeon/DGM-TF#1]
Hi, thank you for your great work! And I have some questions and need your help.

I found that in your implementation, the encoder and decoder have different structures compared to the original DGM paper. Basically, it has 2 conv layers followed by a pooling layer every time. Is it because the mnist data have relatively small sizes and don't need too many layers?

I'm currently trying to use DGM to generate residue maps on CXR images with the size of 224*224(same as in the paper). And I've modified the depth of the model and in/out channels. It has successfully output the parameters for flow 1, 2 and 3. But when it starts training, it has the following bugs.
![9061625328340_ pic_hd](https://user-images.githubusercontent.com/41676649/124360311-a1d29800-dc5b-11eb-9c60-172f7136984f.jpg)
In the encoder, we have modified it to this:
![截屏2021-07-04 上午12 07 11](https://user-images.githubusercontent.com/41676649/124360336-cb8bbf00-dc5b-11eb-989b-4bda2f324978.png)
And in the decoder, we have modified to this:
![截屏2021-07-04 上午12 09 37](https://user-images.githubusercontent.com/41676649/124360398-22919400-dc5c-11eb-9307-041b1f06784b.png)
I really like your work and want to see how it works on my customized dataset. Could you give me some instructions on modifying the model? I'm really a novice to this.
I'll appreciate your response.

**Comment by YeongHyeon** [comment]
Hi @Jacobieee ,

First of all, thank you for your interest.
Basically, the source code given in this repository is constructed in a simplified form with the MNIST dataset.
Thus, the depth of the layer is shallow.

However, the source code does not need additional customizing for applying as your case (CXR).
To reduce the spatial dimension via convolutional operation, I recommend you adjust the parameter 'depth'.
For example, adjust 'depth' from 3 to 5 or other deeper numbers. 
``` python
def __build_model(self, x, ksize=3, verbose=True):

        print("\n-*-*- Flow 1 -*-*-")
        # Module 1: E_G
        self.variables['c_z'] = \
            self.__encoder(x=x, ksize=ksize, outdim=self.zdim, reuse=False, \
            name='enc_g', norm=True, depth=5, verbose=verbose)
        # Module 2: D_G
        self.variables['y_hat'] = \
            self.__decoder(z=self.variables['c_z'], ksize=ksize, reuse=False, \
            name='gen_g', norm=True, depth=5, verbose=verbose)
        # Module 3: D
        self.variables['d_real'] = \
            self.__encoder(x=x, ksize=ksize, outdim=1, reuse=False, \
            name='dis_g', norm=True, depth=5, verbose=verbose)
        # Revisit D
        self.variables['d_fake'] = \
            self.__encoder(x=self.variables['y_hat'], ksize=ksize, outdim=1, reuse=True, \
            name='dis_g', norm=True, depth=5, verbose=False)

        print("\n-*-*- Flow 2 -*-*-")
        # Module 4: E_F
        self.variables['c_s'] = \
            self.__encoder(x=x, ksize=ksize, outdim=self.zdim, reuse=False, \
            name='enc_f', norm=False, depth=5, verbose=verbose)
        # Module 5: D_F
        self.variables['a'] = \
            self.__decoder(z=self.variables['c_s'], ksize=ksize, reuse=False, \
            name='gen_f', norm=True, depth=5, verbose=verbose)

        print("\n-*-*- Flow 3 -*-*-")
        # Vector joint
        self.variables['c_joint'] = \
            tf.concat([self.variables['c_z'], self.variables['c_s']], 1)
        # Module 6: D_J
        self.variables['z_phat'] = \
            self.__decoder(z=self.variables['c_joint'], ksize=ksize, reuse=False, \
            name='gen_j', norm=True, depth=5, verbose=verbose)
```

``` python
def __encoder(self, x, ksize=3, outdim=1, reuse=False, \
        name='enc', activation='relu', norm=True, depth=3, verbose=True):
       ...

def __decoder(self, z, ksize=3, reuse=False, \
        name='dec', activation='relu', norm=True, depth=3, verbose=True):
       ...
```

Please try the above instructions first and let me know if you need more support.
Always good luck to you.
:)

**Comment by Jacobieee** [comment]
Hi @YeongHyeon ,

Thank you so much for your reply.
I have tried your suggestion and during these days, I have also reconstructed the network to make it very similar to that in the paper.
I think both of them work fine but with a little problem.

While I was running your code on mnist, I found that during training, it keeps occupying more RAM. I'm using google colab pro with 25G RAM and after about 70 epochs, it takes all my RAM and I had to stop training.

And this happens much earlier when I train it on my 224*224 images (can only train 8 epochs). I spent a lot of time finding why it keeps eating RAM while training but got nothing. 

Have you also found this issue while implementing it? Could you give me some instructions to tackle this issue? Currently, I don't know if making a data generator would help.
I'd appreciate you so much for your help.

**Comment by YeongHyeon** [comment]
Hi @Jacobieee ,

Thanks for checking the details.
In my case, I used a local GPU server and there was no such problem.
However, I will check it in Colab to confirm and solve the problem which you informing me.

I support your research.
Have a nice day!

**Comment by Jacobieee** [comment]
Hi @YeongHyeon 

I appreciate it so much that you support me!

Today, I found that while running with a larger batch size (128 for mnist), the training will be much faster and the speed of eating RAM will be slower. And I've also found that the RAM occupation increases every time (about 3G) when starting a new epoch. 

Hope the information above will help you solve the issue. And at the same time, I'll also try to solve this issue. 

Again, thank you so much for helping me.

**Comment by Jacobieee** [comment]
Hi @YeongHyeon 

Currently, we have found some pieces of code where there could be a memory leak by printing out the memory change. We found that in function "step" in neuralnet.py [here](https://github.com/YeongHyeon/DGM-TF/blob/ec42e2748ef2aa6d12e6a92a77562e23adbcccfb/source/neuralnet.py#L52), the RAM occupation slowly increases. And it will increase averagely 8G each epoch. 

Hope it's helpful. 

Thanks.
